staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -762,14 +762,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	consistentReadFromStorage := resourceVersion == "" && !(consistentListFromCacheEnabled && requestWatchProgressSupported)
 	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := len(pred.Continue) > 0
-	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
-	hasLimit := pred.Limit > 0 && resourceVersion != "0"
+	// Serve paginated requests about revision "0" and "" (when consistent read from cache is supported)
+	// from watch cache to avoid overwhelming etcd.
+	consistentListFromCache := resourceVersion == "" && (consistentListFromCacheEnabled && requestWatchProgressSupported)
+	hasLimit := pred.Limit > 0 && !(resourceVersion == "0" || consistentListFromCache)
```

Comment:
Is this even needed at this point?

Given that you already implemented the support for limit in watchcache, it just works in watchcache. So existence of limit is no longer a reason to send request to etcd.
So can't we just remove the `hasLimit` completely?

Reply:
I thought that we want to apply the limit only to requests that request a consistent snapshot of data from the cache as described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache#what-if-the-watch-cache-is-stale

If that is not the case then yeah we could get rid of `hasLimit` and always serve the first page from the cache.

Reply:
That was my original intention, but thinking more about that, I think it doesn't make much sense.
In order to handle most of list requests (including those from kubectl) we need to support limit in watchcache.

For cases where we already handle a given semantic with limit unset in watchcache, as soon as we support the limit, it just doesn't make any sense to continue sending them to etcd.

So I'm treating this PR as somewhat orthogonal:
- implement "limit" in watchcache [the implementation is fairly trivial]
- given we have it, use it everywhere we have [to amplify the impact of ConsistentListsFromCache]

Reply:
okay, sgtm, i will update the pr, it should require much changes.